google/api_core/retry_streaming.py

```diff
@@ -0,0 +1,220 @@
+        target: The generator function to call and retry. This must be a
+            nullary function - apply arguments with `functools.partial`.
+        predicate: A callable used to determine if an
+            exception raised by the target should be considered retryable.
+            It should return True to retry or False otherwise.
+        sleep_generator: An infinite iterator that determines
+            how long to sleep between retries.
+        timeout: How long to keep retrying the target.
+            Note: timeout is only checked before initiating a retry, so the target may
+            run past the timeout value as long as it is healthy.
+        on_error: A function to call while processing a
+            retryable exception.  Any error raised by this function will *not*
+            be caught.
+        exception_factory: A function that is called when the retryable reaches
+            a terminal failure state, used to construct an exception to be raised.
+            It it given a list of all exceptions encountered, a boolean indicating
+            whether the failure was due to a timeout, and the original timeout value
+            as arguments. It should return a tuple of the exception to be raised,
+            along with the cause exception if any.
+            If not provided, a default implementation will raise a RetryError
+            on timeout, or the last exception encountered otherwise.
+
+    Returns:
+        Generator: A retryable generator that wraps the target generator function.
+
+    Raises:
+        ValueError: If the sleep generator stops yielding values.
+        Exception: a custom exception specified by the exception_factory if provided.
+            If no exception_factory is provided:
+                google.api_core.RetryError: If the deadline is exceeded while retrying.
+                Exception: If the target raises an error that isn't retryable.
+    """
+
+    timeout = kwargs.get("deadline", timeout)
+    deadline: Optional[float] = time.monotonic() + timeout if timeout else None
+    error_list: List[Exception] = []
+    exc_factory = partial(
+        exception_factory or _build_timeout_error, timeout_val=timeout
+    )
+
+    for sleep in sleep_generator:
+        # Start a new retry loop
+        try:
+            # create and yeild from a new instance of the generator from input generator function
+            subgenerator = target()
+            return (yield from subgenerator)
+        # handle exceptions raised by the subgenerator
+        except Exception as exc:
+            error_list.append(exc)
+            if not predicate(exc):
+                final_exc, source_exc = exc_factory(
+                    exc_list=error_list, is_timeout=False
+                )
+                raise final_exc from source_exc
+            if on_error is not None:
+                on_error(exc)
+        finally:
+            if subgenerator is not None and getattr(subgenerator, "close", None):
```

Comment:
Good catch. I believe this is left-over from an earlier attempt, where I was trying to implement more of the generator protocol manually instead of relying on `yield from` (I  originally wanted to implement timeouts in between yields as well)

I think it should be safe to remove this extra check, and the unit tests seem to back that up